### ISSUE TITLE
Can't open file "/tmp/a9mCPkqcRm": Too many open files at

### ISSUE THREAD

**Issue by Htbaa** [majensen/rest-neo4p#7]
Hi, another issue I'm running into. I noticed that many temporary files are created, because they're not cleaned up after. Resulting in the error message from the title. The context I'm running REST::Neo4p in is a Mojolicious webapplicatie running behind Hypnotoad and the Minion job worker. I'm not using REST::Neo4p directly but through DBD::Neo4j.

I'm not calling `$sth->finish()` because I remember reading that's not something you're supposed to do...

**Comment by Htbaa** [comment]
Hmm, calling finish on `$sth` seems to do the trick. So I guess that's something that should be documented in `DBD::Neo4p` as it's not directly related to `REST::Neo4p`?

**Comment by majensen** [comment]
thanks @Htbaa - I think this is an issue; I will look into it. (I remember the 'finish' hint too, probably means DBD should do it naturally somewhere.)

**Comment by Htbaa** [comment]
Somehow I can't seem to find that mention of using finish in the documentation. Maybe I read it in an online article?

I was wondering, are the temporary files really required? Even when using finish I'm still experiencing lots of open files and I made sure finish is being called after (fetching after) execute.

**Comment by Htbaa** [comment]
The temporary files really are starting to become a hindrance for me. Can't we just put this stuff in memory instead of temporary files? Even when calling `finish` file handles aren't instantly closed so the problem persists.

**Comment by Htbaa** [comment]
Sorry to keep bothering you with this. But I was wondering, what about simply deleting the temporary file itself by calling `unlink` on it in `REST::Neo4p::Query->finish()`? And additionally to fix the issue in `DBD::Neo4p` maybe call `$sth->{"${prefix}_query_obj"}->finish` instead of undeffing it there?

By unlinking the temporary file ourselves we don't have to wait for the garbage collector doing so.

**Comment by majensen** [comment]
Can you give me a toy example that has the behavior you're seeing? I think you're right, but it seems like there is a fundamental smelliness about the code. For example, maybe one fetches one or a few rows, but not all rows, and if all rows are not fetched, then a ref isn't properly handled (and a file isn't closed out).

**Comment by Htbaa** [comment]
I'll try to fix up an example. Long running processes such as background workers have this issue when executing lots of calls to the Neo4j API. One specific example is having to apply a mutation to all nodes with a certain label. E.g. best example I can give is having 4600+ nodes for which I needed to add a property.

```
sub slugify {
    my ($job) = @_;
    my $neo4j = $job->app->neo4j;
    try {
        my $sth_update = $neo4j->prepare(
            q|
                    MATCH (a:Actor {name: {name} })
                    SET a.slug = {slug}
                |
        );
        my $sth = $neo4j->prepare(
            q|
                    MATCH (a:Actor)
                    WHERE NOT EXISTS(a.slug)
                    RETURN a.name
                |
        );
        $sth->execute();
        while (my $row = $sth->fetch()) {
            my $slug = $job->app->slugify($row->[0]);
            $sth_update->execute($row->[0], $slug);
            $sth_update->finish();
        }
        $sth->finish();
    }
    catch {
        $job->fail($_);
    }
    finally {
        $neo4j->disconnect;
    };
}
```

This is an excerpt of a background job. I'm using `DBD::Neo4p` here. This would normally end up in the stated error message. After I applied the patches from both pull requests it didn't.

**Comment by Htbaa** [comment]
I must add that with a background job like this the limit is reached fast. But even in my Mojolicious app where I'm executing far less queries it happens as time goes by. And I call `finish` on all statements plus they go out of scope after a while, but not fast enough to clean up it seems.

Aside from the provided pull requests the way I'm fixing it now is to restart my application server every 15 minutes (I'm very thankful for Hypnotoad's hot deploy feature) and my job queue (Minion) as well. Not an ideal solution for something that can be fixed by deleting the temp files.

**Comment by Htbaa** [comment]
This has been resolved with pull request #8 .
